edb/edgeql/compiler/casts.py

```diff
@@ -236,6 +241,20 @@ def compile_cast(
     )
 
 
+def _has_common_concrete_scalar(
+        orig_stype: s_types.Type,
+        new_stype: s_types.Type, *,
+        ctx: context.ContextLevel) -> bool:
+    return bool(
+        isinstance(orig_stype, s_scalars.ScalarType)
+        and isinstance(new_stype, s_scalars.ScalarType)
+        and (nearest := s_utils.get_class_nearest_common_ancestors(
+            ctx.env.schema, [new_stype, orig_stype]))
+        and len(nearest) == 1
+        and not nearest[0].get_abstract(ctx.env.schema)
```

Comment:
This wouldn't work if the nearest common ancestor is an abstract subtype of a non-abstract parent, e.g

```
create abstract scalar type foo extending str;
create scalar type bar1 extending foo;
create scalar type bar2 extending foo;
```

So:

```suggestion
        and nearest[0].maybe_get_topmost_concrete_base(ctx.env.schema) is not None
```

Reply:
Good catch

Reply:
Actually, I think the whole check should just compare the results of `maybe_get_topmost_concrete_base`, because

```
create abstract scalar type foo extending str;
create abstract scalar type bar extending str;
create scalar type bar1 extending foo, bar;
create scalar type bar2 extending foo, bar;
```

Reply:
Oh, yeah, indeed; I had misremembered and thought scalars could only have one base, but it's actually they can have one *concrete* ancestor

Reply:
And this is probably faster too